File: php/07_auth_failures.php
<?php
// A07 - Auth Failures. VULN: session fixation + weak remember-me.
// VULN: A07
function login($user) {
    session_start();
    // VULN: accepts session id from client without regeneration
    $_SESSION['user'] = $user;
    setcookie('remember', base64_encode($user . ':password123'), time() + 86400 * 365);
}

File: go/07_auth_failures.go
// A07 - Auth Failures. VULN: constant-time compare missing, weak session.
package main

import "time"

// VULN: A07 - session never expires
type Session struct {
	User string
}

var sessions = map[string]Session{}

func Login(user, pass string) string {
	if user == "admin" && pass == "admin" {
		token := user + string(rune(time.Now().Unix()))
		sessions[token] = Session{User: user}
		return token
	}
	return ""
}
